src/Symfony/Component/DomCrawler/Crawler.php

```diff
@@ -766,17 +766,23 @@ public function extract($attributes)
      * @param string $xpath An XPath expression
      *
      * @return static
+     *
+     * @throws \InvalidArgumentException If contextNode is not null and not part of the document of this crawler
      */
-    public function filterXPath($xpath)
+    public function filterXPath($xpath, \DOMNode $contextNode = null)
     {
+        if (null !== $contextNode && $contextNode->ownerDocument !== $this->document) {
```

Comment:
This is not enough to avoid breaking the API. It could still be a node higher in the document that the ones belonging to the current Crawler instead, which would allow `filterXPath` will a context node to return nodes that could never be reached by a normal `filterXPath` call.
To make this respect the API of the Crawler, you would have to validate that the provided node is either one of the crawler nodes or one of their descendants

Reply:
While I understand if one does not like this kind of API I don't see an issue in this.
The crawler operates on one document and it is possible to add new nodes at any point. To me there is no harm even if nodes not part of the crawler are used as a context node. The crawler returns just sub-nodes of this passed node and operates on data of this document.

Reply:
The fact that subcrawlers that we return operate on the same DOMDocument that the original crawler is purely an implementation detail, and we **don't want** to allow devs to rely on it (Symfony 2 and 3 were not doing that, and were much slower before of that).
The API of the Crawler is that it operates only in the context of its nodes.